src/Control/Director.php

```diff
@@ -373,28 +350,77 @@ protected static function handleRequest(HTTPRequest $request)
-                    // Redirection
-                    $response = new HTTPResponse();
-                    $response->redirect(static::absoluteURL($arguments['Redirect']));
-                    return $response;
+                    $handler = function () use ($arguments) {
+                        // Redirection
+                        $response = new HTTPResponse();
+                        $response->redirect(static::absoluteURL($arguments['Redirect']));
+                        return $response;
+                    };
+                    break;
                 }
 
                 // Find the controller name
                 $controller = $arguments['Controller'];
                 $controllerObj = Injector::inst()->create($controller);
 
-                try {
-                    return $controllerObj->handleRequest($request);
-                } catch (HTTPResponse_Exception $responseException) {
-                    return $responseException->getResponse();
-                }
+                // Handler for calling a controller
+                $handler = function ($request) use ($controllerObj) {
+                    try {
+                        return $controllerObj->handleRequest($request);
+                    } catch (HTTPResponse_Exception $responseException) {
+                        return $responseException->getResponse();
+                    }
+                };
+                break;
             }
         }
 
-        // No URL rules matched, so return a 404 error.
-        return new HTTPResponse('No URL rule was matched', 404);
+        // Call the handler with the given middlewares
+        $response = self::callWithMiddlewares(
+            $request,
+            $middlewares,
+            $handler
+        );
+
+        // Note that if a different request was previously registered, this will now be lost
+        // In these cases it's better to use Kernel::nest() prior to kicking off a nested request
+        Injector::inst()->unregisterNamedObject(HTTPRequest::class);
+
+        return $response;
+    }
+
+    /**
+     * Call the given request handler with the given middlewares
+     * Middlewares are specified as Injector service names
+     *
+     * @param $request The request to pass to the handler
+     * @param $middlewareNames The services names of the middlewares to apply
+     * @param $handler The request handler
+     */
+    protected static function callWithMiddlewares(HTTPRequest $request, array $middlewareNames, callable $handler)
+    {
+        $next = $handler;
+
+        if ($middlewareNames) {
+            $middlewares = array_map(
+                function ($name) {
+                    return Injector::inst()->get($name);
+                },
+                $middlewareNames
+            );
+
+            // Reverse middlewares
+            /** @var HTTPMiddleware $middleware */
+            foreach (array_reverse($middlewares) as $middleware) {
```

Comment:
Ok, you'll need to add another option to allow custom extra middlewares to be passed in for controller-specific middleware.